File: runtime/go/cmd/pebble-control/skills.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"sort"
	"strings"
)

type skillGuideResult struct {
	Name                string `json:"name"`
	RequestedAppVersion string `json:"requestedAppVersion"`
	ResolvedAppVersion  string `json:"resolvedAppVersion"`
	ReleaseRevision     int    `json:"releaseRevision"`
	FallbackReason      string `json:"fallbackReason,omitempty"`
	Content             string `json:"content"`
}

func runSkills(args []string, output io.Writer) error {
	if len(args) == 0 || args[0] != "get" {
		return fmt.Errorf("usage: pebble skills get [--app-version VERSION] [--json] <skill-name>")
	}
	flags := flag.NewFlagSet("skills get", flag.ContinueOnError)
	flags.SetOutput(io.Discard)
	appVersion := flags.String("app-version", bundledSkillAppVersion, "Pebble application version")
	jsonOutput := flags.Bool("json", false, "print structured JSON")
	if err := flags.Parse(args[1:]); err != nil {
		return err
	}
	if flags.NArg() != 1 {
		return fmt.Errorf("skills get requires exactly one skill name")
	}
	name := strings.TrimSpace(flags.Arg(0))
	guide, found := bundledSkillGuides[name]
	if !found {
		names := make([]string, 0, len(bundledSkillGuides))
		for candidate := range bundledSkillGuides {
			names = append(names, candidate)
		}
		sort.Strings(names)
		return fmt.Errorf("unknown skill %q; available skills: %s", name, strings.Join(names, ", "))
	}
	requestedVersion := strings.TrimSpace(*appVersion)
	if requestedVersion == "" {
		requestedVersion = bundledSkillAppVersion
	}
	result := skillGuideResult{
		Name: name, RequestedAppVersion: requestedVersion,
		ResolvedAppVersion: bundledSkillAppVersion, ReleaseRevision: guide.Revision,
		Content: guide.Content,
	}
	if requestedVersion != bundledSkillAppVersion {
		// Why: a bundled guide is safer than guessing commands when a development
		// build has no explicit historical mapping yet.
		result.FallbackReason = "requested version is unmapped; using the current bundled guide"
	}
	if *jsonOutput {
		return json.NewEncoder(output).Encode(result)
	}
	_, err := io.WriteString(output, result.Content)
	return err
}

File: packages/product-core/cli/handlers/skills.ts
import { readFile, stat } from 'node:fs/promises'
import path from 'node:path'

import type { CommandHandler } from '../dispatch'
import { RuntimeClientError } from '../runtime-client'

type SkillGuideResult = {
  name: string
  requestedAppVersion: string
  resolvedAppVersion: string
  releaseRevision: number
  fallbackReason?: string
  content: string
}

async function isPebbleRoot(candidate: string): Promise<boolean> {
  const paths = [path.join(candidate, 'package.json'), path.join(candidate, 'skill-guides')]
  const results = await Promise.all(paths.map((entry) => stat(entry).catch(() => null)))
  return results[0]?.isFile() === true && results[1]?.isDirectory() === true
}

export async function findBundledSkillRoot(cwd: string): Promise<string> {
  const starts = [cwd, __dirname]
  for (const start of starts) {
    let candidate = path.resolve(start)
    while (true) {
      if (await isPebbleRoot(candidate)) {
        return candidate
      }
      const parent = path.dirname(candidate)
      if (parent === candidate) {
        break
      }
      candidate = parent
    }
  }
  throw new RuntimeClientError(
    'not_found',
    'Could not locate bundled Pebble skill guides from this development CLI.'
  )
}

async function readSkillGuide(cwd: string, name: string, requestedVersion?: string) {
  const root = await findBundledSkillRoot(cwd)
  const manifest = JSON.parse(
    await readFile(path.join(root, 'resources/skills/current-manifest.json'), 'utf8')
  ) as {
    appVersion: string
    skills: { name: string; releaseRevision: number }[]
  }
  const entry = manifest.skills.find((skill) => skill.name === name)
  if (!entry) {
    const available = manifest.skills
      .map((skill) => skill.name)
      .sort()
      .join(', ')
    throw new RuntimeClientError(
      'invalid_argument',
      `Unknown skill: ${name}. Available skills: ${available}`
    )
  }
  const content = await readFile(path.join(root, 'skill-guides', `${name}.md`), 'utf8')
  const requestedAppVersion = requestedVersion?.trim() || manifest.appVersion
  return {
    name,
    requestedAppVersion,
    resolvedAppVersion: manifest.appVersion,
    releaseRevision: entry.releaseRevision,
    ...(requestedAppVersion === manifest.appVersion
      ? {}
      : { fallbackReason: 'requested version is unmapped; using the current bundled guide' }),
    content
  } satisfies SkillGuideResult
}

export const SKILL_HANDLERS: Record<string, CommandHandler> = {
  'skills get': async ({ flags, cwd, json }) => {
    const name = flags.get('name')
    if (typeof name !== 'string' || name.trim() === '') {
      throw new RuntimeClientError('invalid_argument', 'skills get requires one skill name')
    }
    const appVersion = flags.get('app-version')
    const result = await readSkillGuide(
      cwd,
      name.trim(),
      typeof appVersion === 'string' ? appVersion : undefined
    )
    process.stdout.write(json ? `${JSON.stringify(result)}\n` : result.content)
  }
}
